src/rust/catnap/socket.rs

```diff
@@ -331,14 +181,235 @@ impl Socket {
+        match op {
+            SocketOp::Bind => Err(fail(
+                op,
+                &(format!("socket is connecting to address: {:?}", self.remote)),
+                libc::EINVAL,
+            )),
+            SocketOp::Listen => Err(fail(
+                op,
+                &(format!("socket is connecting to address: {:?}", self.remote)),
+                libc::EADDRINUSE,
+            )),
+            SocketOp::Accept => Err(fail(
+                op,
+                &(format!("socket is connecting to address: {:?}", self.remote)),
+                libc::EINVAL,
+            )),
+            SocketOp::Accepted => Err(fail(
+                op,
+                &(format!("socket is connecting to address: {:?}", self.remote)),
+                libc::EINVAL,
+            )),
+            SocketOp::Connect => Err(fail(
+                op,
+                &(format!("socket already is connecting to address: {:?}", self.remote)),
+                libc::EINPROGRESS,
+            )),
+            SocketOp::Connected => Ok(SocketState::Connected),
+            SocketOp::Close => Ok(SocketState::Closing),
+            SocketOp::Closed => Err(fail(op, &(format!("socket is busy")), libc::EBUSY)),
+        }
+    }
+
+    fn connected_state(&self, op: SocketOp) -> Result<SocketState, Fail> {
+        match op {
+            // Does this make sense if we didn't go through the Connecting state?
+            SocketOp::Bind => Ok(SocketState::Connected),
+            SocketOp::Listen => Err(fail(
+                op,
+                &(format!("socket is already connected to address: {:?}", self.remote)),
+                libc::EISCONN,
+            )),
+            SocketOp::Accept => Err(fail(
+                op,
+                &(format!("socket is already connected to address: {:?}", self.remote)),
+                libc::EISCONN,
+            )),
+            SocketOp::Accepted => Err(fail(
+                op,
+                &(format!("socket is already connected to address: {:?}", self.remote)),
+                libc::EISCONN,
+            )),
+            SocketOp::Connect => Err(fail(
+                op,
+                &(format!("socket is connected to address: {:?}", self.remote)),
+                libc::EISCONN,
+            )),
+            // Should this be possible without going through the Connecting state?
+            SocketOp::Connected => Err(fail(
+                op,
+                &(format!("socket is connected to address: {:?}", self.remote)),
+                libc::EISCONN,
+            )),
+            SocketOp::Close => Ok(SocketState::Closing),
+            SocketOp::Closed => Err(fail(op, &(format!("socket is busy")), libc::EBUSY)),
+        }
+    }
+
+    fn closing_state(&self, op: SocketOp) -> Result<SocketState, Fail> {
+        match op {
+            SocketOp::Bind => Err(fail(op, &(format!("socket is closing")), libc::EBADF)),
```

Comment:
It would be good to evaluate if moving this to a common place (variable/function) would be better.

Reply:
Yes, we might use a macro eventually, but I'm not sure it would make it much shorter. We could try to shorten the error messages themselves instead if we want it to be more compact